File: trunk/python/cohorte.python/cohorte/remote/signals_discovery.py
#!/usr/bin/env python
# -- Content-Encoding: UTF-8 --
"""
COHORTE Remote Services: Discovery based on Signals

:author: Thomas Calmant
:copyright: Copyright 2014, isandlaTech
:license: Apache License 2.0
:version: 1.1
:status: Beta

..

    Copyright 2013 isandlaTech

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
"""

# Documentation strings format
__docformat__ = "restructuredtext en"

# Module version
__version_info__ = (1, 1, 0)
__version__ = ".".join(str(x) for x in __version_info__)

# ------------------------------------------------------------------------------

# Cohorte
import cohorte.signals

# iPOPO Decorators
from pelix.ipopo.decorators import ComponentFactory, Provides, Validate, \
    Invalidate, Requires
import pelix.remote
import pelix.remote.beans as beans

# Standard library
import logging

# ------------------------------------------------------------------------------

SIGNALS_NAME_PREFIX = "/cohorte/remote-service-broadcaster"
SIGNALS_MATCH_ALL = "{0}/*".format(SIGNALS_NAME_PREFIX)
SIGNAL_REMOTE_EVENT = "{0}/remote-event".format(SIGNALS_NAME_PREFIX)
SIGNAL_REQUEST_ENDPOINTS = "{0}/request-endpoints".format(SIGNALS_NAME_PREFIX)

# ------------------------------------------------------------------------------

_logger = logging.getLogger(__name__)

# ------------------------------------------------------------------------------

ENDPOINT_EVENT_TYPE = "org.cohorte.remote.discovery.signals.EEndpointEventType"
EVENT_REGISTERED = {'javaClass': ENDPOINT_EVENT_TYPE, 'enumValue': "REGISTERED"}
EVENT_MODIFIED = {'javaClass': ENDPOINT_EVENT_TYPE, 'enumValue': "MODIFIED"}
EVENT_UNREGISTERED = {'javaClass': ENDPOINT_EVENT_TYPE,
                      'enumValue': "UNREGISTERED"}

class EndpointEventBean(object):
    """
    An endpoint event bean
    """
    javaClass = "org.cohorte.remote.discovery.signals.EndpointEventBean"
    """ Java class (for Jabsorb) """

    def __init__(self, event_type):
        """
        Sets up members
        """
        self.endpoints = tuple()
        self.type = event_type


    def set_endpoints(self, export_endpoints):
        """
        Make Java beans from endpoints
        """
        self.endpoints = tuple(EndpointDescriptionBean(export_endpoint)
                               for export_endpoint in export_endpoints)

class EndpointDescriptionBean(object):
    """
    An endpoint description bean
    """
    javaClass = "org.cohorte.remote.discovery.signals.EndpointDescriptionBean"
    """ Java class (for Jabsorb) """

    def __init__(self, export_endpoint):
        """
        Sets up the bean according to the given ExportEndpoint bean

        :param export_endpoint: An ExportEndpoint bean
        """
        self.uid = export_endpoint.uid
        self.frameworkUid = export_endpoint.framework
        self.name = export_endpoint.name

        # Use tuple to force arrays in the Jabsorb serialization
        self.configurations = tuple(export_endpoint.configurations)
        self.specifications = tuple(export_endpoint.specifications)

        # Store merged properties
        self.properties = export_endpoint.make_import_properties()

# ------------------------------------------------------------------------------

@ComponentFactory("cohorte-remote-discovery-signals-factory")
@Provides((pelix.remote.SERVICE_EXPORT_ENDPOINT_LISTENER,
           cohorte.signals.SERVICE_ISOLATE_PRESENCE_LISTENER))
@Requires('_dispatcher', pelix.remote.SERVICE_DISPATCHER)
@Requires("_registry", pelix.remote.SERVICE_REGISTRY)
@Requires("_directory", cohorte.SERVICE_SIGNALS_DIRECTORY)
@Requires("_receiver", cohorte.SERVICE_SIGNALS_RECEIVER)
@Requires("_sender", cohorte.SERVICE_SIGNALS_SENDER)
class SignalsDiscovery(object):
    """
    Remote services discovery and notification using COHORTE Signals
    """
    def __init__(self):
        """
        Sets up the component
        """
        # Bundle context
        self._context = None

        # End points registry
        self._dispatcher = None
        self._registry = None

        # Signals
        self._directory = None
        self._receiver = None
        self._sender = None


    def _request_endpoints(self, isolate=None):
        """
        Requests the services exported by the given isolate. If isolate is None,
        then the request is sent to all known isolates.

        :param: isolate: An isolate UID (optional)
        """
        # Prepare our event, to be sent with the request
        local_event = self._make_exports_event()

        if not isolate:
            raw_results = self._sender.send(SIGNAL_REQUEST_ENDPOINTS,
                                            local_event,
                                        dir_group=cohorte.signals.GROUP_OTHERS)

        else:
            raw_results = self._sender.send(SIGNAL_REQUEST_ENDPOINTS,
                                            local_event, isolate=isolate)

        if raw_results is None:
            # Nothing to do
            return

        else:
            # Extract information
            sig_results = raw_results[0]
            if not sig_results:
                # Nothing to do...
                return

        for isolate_uid, isolate_sigresult in sig_results.items():
            # Get the first result as the event
            if not isolate_sigresult:
                _logger.warning("Isolate %s returned None", isolate_uid)
                continue

            isolate_results = isolate_sigresult['results']
            if not isolate_results:
                _logger.warning("Isolate %s returned nothing", isolate_uid)
                continue

            event = isolate_results[0]

            # Get the node of the isolate
            node_uid = self._directory.get_isolate_node(isolate_uid)

            # Handle the event
            self.__handle_event(event,
                                self._directory.get_host_for_node(node_uid))


    def _make_exports_event(self):
        """
        Returns a single REGISTERED endpoint event, matching the endpoints
        exported by the local isolate
        """
        # Get the export endpoints
        event = EndpointEventBean(EVENT_REGISTERED)

        event.set_endpoints(self._dispatcher.get_endpoints())
        return event


    def _send_event(self, event):
        """
        Sends an event using Signals.

        :param event: An EndpointEventBean
        """
        self._sender.fire(SIGNAL_REMOTE_EVENT, event,
                          dir_group=cohorte.signals.GROUP_OTHERS)


    def endpoints_added(self, endpoints):
        """
        New endpoints are exported
        """
        # Prepare the event bean
        event = EndpointEventBean(EVENT_REGISTERED)
        event.set_endpoints(endpoints)

        # Send the signal
        self._send_event(event)


    def endpoint_updated(self, endpoint, old_properties):
        """
        An end point is updated
        """
        # Prepare the event bean
        event = EndpointEventBean(EVENT_MODIFIED)
        event.set_endpoints([endpoint])

        # Send the signal
        self._send_event(event)


    def endpoint_removed(self, endpoint):
        """
        An end point has been removed

        :param endpoint: An ExportEndpoint bean
        """
        # Prepare the event bean
        event = EndpointEventBean(EVENT_UNREGISTERED)
        event.set_endpoints([endpoint])

        # Send the signal
        self._send_event(event)


    def __handle_event(self, event, sender_address):
        """
        Hanldles an endpoint event

        :param event: An EndpointEvent bean
        :param sender_address: The address of the event sender
        """
        kind = event.type

        if kind == EVENT_REGISTERED:
            # Registration of endpoints
            self.__register_endpoints(event.endpoints, sender_address)

        elif kind == EVENT_MODIFIED:
            # Single endpoint updated
            endpoint = event.endpoints[0]
            self._registry.update(endpoint.uid, endpoint.properties)

        elif kind == EVENT_UNREGISTERED:
            # Single endpoint unregistered
            endpoint = event.endpoints[0]
            self._registry.remove(endpoint.uid)

        else:
            # Unknown kind of event
            _logger.warning("Unknown RS event type: %s", kind)


    def __register_endpoints(self, endpoints, server):
        """
        Registers the new endpoints
        """
        for endpoint in endpoints:
            # Make the ImportEndpoint bean
            import_endpoint = beans.ImportEndpoint(endpoint['uid'],
                                                   endpoint['frameworkUid'],
                                                   endpoint['configurations'],
                                                   endpoint['name'],
                                                   endpoint['specifications'],
                                                   endpoint['properties'])
            import_endpoint.server = server

            # Notify the import registry
            self._registry.add(import_endpoint)


    def handle_isolate_presence(self, uid, name, node, event):
        """
        Handles an isolate presence event

        :param uid: UID of the isolate
        :param name: Name of the isolate
        :param node: Node of the isolate
        :param event: Kind of event
        """
        if event == cohorte.signals.ISOLATE_REGISTERED:
            # New isolate detected
            self._request_endpoints(uid)

        elif event == cohorte.signals.ISOLATE_UNREGISTERED:
            # Isolate lost
            self._registry.lost_framework(uid)


    def handle_received_signal(self, name, signal_data):
        """
        Called when a remote services signal is received
        """
        # Get the sender address
        sender = signal_data['senderAddress']

        if name == SIGNAL_REMOTE_EVENT:
            # Received a remote service event
            data = signal_data["signalContent"]

            # Single event
            self.__handle_event(data, sender)

        elif name == SIGNAL_REQUEST_ENDPOINTS:
            # An isolate requests all of our exported services
            event = signal_data["signalContent"]

            # Register remote endpoints
            self.__handle_event(event, sender)

            # Return our endpoints
            return self._make_exports_event()


    @Invalidate
    def invalidate(self, context):
        """
        Component invalidated
        """
        # Unregister from signals
        self._receiver.unregister_listener(SIGNALS_MATCH_ALL, self)

        self._context = None
        _logger.info("Signals discovery invalidated")


    @Validate
    def validate(self, context):
        """
        Component validated
        """
        self._context = context

        # Register to signals
        self._receiver.register_listener(SIGNALS_MATCH_ALL, self)

        # Send a discovery request
        self._request_endpoints()

        _logger.info("Signals discovery validated")

File: trunk/java/remote-services/org.cohorte.remote.discovery.signals/src/org/cohorte/remote/discovery/signals/SignalsDiscovery.java
/**
 * Copyright 2014 isandlaTech
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.cohorte.remote.discovery.signals;

import java.util.Map;
import java.util.Map.Entry;

import org.apache.felix.ipojo.annotations.Component;
import org.apache.felix.ipojo.annotations.Instantiate;
import org.apache.felix.ipojo.annotations.Invalidate;
import org.apache.felix.ipojo.annotations.Provides;
import org.apache.felix.ipojo.annotations.Requires;
import org.apache.felix.ipojo.annotations.Validate;
import org.cohorte.remote.ExportEndpoint;
import org.cohorte.remote.IExportEndpointListener;
import org.cohorte.remote.IExportsDispatcher;
import org.cohorte.remote.IImportsRegistry;
import org.cohorte.remote.ImportEndpoint;
import org.osgi.service.log.LogService;
import org.psem2m.isolates.constants.ISignalsConstants;
import org.psem2m.isolates.services.monitoring.IIsolatePresenceListener;
import org.psem2m.signals.ISignalBroadcaster;
import org.psem2m.signals.ISignalData;
import org.psem2m.signals.ISignalDirectory;
import org.psem2m.signals.ISignalDirectory.EBaseGroup;
import org.psem2m.signals.ISignalListener;
import org.psem2m.signals.ISignalReceiver;
import org.psem2m.signals.ISignalSendResult;

/**
 * Remote services discovery based on COHORTE Remote Services
 * 
 * @author Thomas Calmant
 */
@Component(name = "cohorte-remote-discovery-signals-factory")
@Provides(specifications = { IExportEndpointListener.class,
        IIsolatePresenceListener.class })
@Instantiate(name = "cohorte-remote-discovery-signals")
public class SignalsDiscovery implements IExportEndpointListener,
        ISignalListener, IIsolatePresenceListener {

    /** Signals directory */
    @Requires
    private ISignalDirectory pDirectory;

    /** Exported endpoints dispatcher */
    @Requires
    private IExportsDispatcher pDispatcher;

    /** The logger */
    @Requires
    private LogService pLogger;

    /** Signals receiver */
    @Requires
    private ISignalReceiver pReceiver;

    /** Imported services registry */
    @Requires
    private IImportsRegistry pRegistry;

    /** Signals sender */
    @Requires
    private ISignalBroadcaster pSender;

    /*
     * (non-Javadoc)
     * 
     * @see
     * org.cohorte.remote.pelix.IExportEndpointListener#endpointRemoved(org.
     * cohorte.remote.pelix.ExportEndpoint)
     */
    @Override
    public void endpointRemoved(final ExportEndpoint aEndpoint) {

        // Prepare the event bean
        final EndpointEventBean event = new EndpointEventBean(
                EEndpointEventType.UNREGISTERED, aEndpoint);

        // Send it
        sendEvent(event);
    }

    /*
     * (non-Javadoc)
     * 
     * @see
     * org.cohorte.remote.pelix.IExportEndpointListener#endpointsAdded(org.cohorte
     * .remote.pelix.ExportEndpoint[])
     */
    @Override
    public void endpointsAdded(final ExportEndpoint[] aEndpoints) {

        // Prepare the event bean
        final EndpointEventBean event = new EndpointEventBean(
                EEndpointEventType.REGISTERED, aEndpoints);

        // Send it
        sendEvent(event);
    }

    /*
     * (non-Javadoc)
     * 
     * @see
     * org.cohorte.remote.pelix.IExportEndpointListener#endpointUpdated(org.
     * cohorte.remote.pelix.ExportEndpoint, java.util.Map)
     */
    @Override
    public void endpointUpdated(final ExportEndpoint aEndpoint,
            final Map<String, Object> aOldProperties) {

        // Prepare the event bean
        final EndpointEventBean event = new EndpointEventBean(
                EEndpointEventType.MODIFIED, aEndpoint);

        // Send it
        sendEvent(event);
    }

    /**
     * Handles an endpoint event
     * 
     * @param aEvent
     *            An endpoint event
     * @param aSenderAddress
     *            The address of the event sender
     */
    private void handleEvent(final EndpointEventBean aEvent,
            final String aSenderAddress) {

        final EndpointDescriptionBean[] endpoints = aEvent.getEndpoints();

        switch (aEvent.getType()) {
        case REGISTERED:
            // Registration of endpoints
            registerEndpoints(endpoints, aSenderAddress);
            break;

        case MODIFIED: {
            // Single endpoint updated
            final EndpointDescriptionBean endpoint = endpoints[0];
            pRegistry.update(endpoint.getUid(), endpoint.getProperties());
            break;
        }

        case UNREGISTERED: {
            // Single endpoint unregistered
            final EndpointDescriptionBean endpoint = endpoints[0];
            pRegistry.remove(endpoint.getUid());
            break;
        }
        }
    }

    /*
     * (non-Javadoc)
     * 
     * @see
     * org.psem2m.signals.ISignalListener#handleReceivedSignal(java.lang.String,
     * org.psem2m.signals.ISignalData)
     */
    @Override
    public Object handleReceivedSignal(final String aSignalName,
            final ISignalData aSignalData) {

        if (ISignalsConstants.BROADCASTER_SIGNAL_REMOTE_EVENT
                .equals(aSignalName)) {
            // Get the event
            final EndpointEventBean event = (EndpointEventBean) aSignalData
                    .getSignalContent();
            handleEvent(event, aSignalData.getSenderAddress());

            // Return nothing
            return null;

        } else if (ISignalsConstants.BROADCASTER_SIGNAL_REQUEST_ENDPOINTS
                .equals(aSignalName)) {
            // Get the events
            final EndpointEventBean event = (EndpointEventBean) aSignalData
                    .getSignalContent();

            // Register remove endpoints
            handleEvent(event, aSignalData.getSenderAddress());

            // Return our endpoints as an event
            return makeExportsEvent();
        }

        // Unhandled signal
        return null;
    }

    /**
     * Component invalidated
     */
    @Invalidate
    public void invalidate() {

        pReceiver.unregisterListener(
                ISignalsConstants.PREFIX_BROADCASTER_SIGNAL_NAME
                        + ISignalsConstants.MATCH_ALL, this);

        pLogger.log(LogService.LOG_INFO, "Signals Discovery gone");
    }

    /*
     * (non-Javadoc)
     * 
     * @see
     * org.psem2m.isolates.services.monitoring.IIsolatePresenceListener#isolateLost
     * (java.lang.String, java.lang.String, java.lang.String)
     */
    @Override
    public void isolateLost(final String aUID, final String aName,
            final String aNodeUID) {

        // Notify the imports registry
        pRegistry.lostFramework(aUID);
    }

    /*
     * (non-Javadoc)
     * 
     * @see
     * org.psem2m.isolates.services.monitoring.IIsolatePresenceListener#isolateReady
     * (java.lang.String, java.lang.String, java.lang.String)
     */
    @Override
    public void isolateReady(final String aIsolateUID, final String aName,
            final String aNodeUID) {

        // Request the endpoints of the given isolate
        requestEndpoints(aIsolateUID);
    }

    /**
     * Returns a signle REGISTERED endpoint event, matching the endpoints
     * exported by the local isolate
     * 
     * @return An endpoint event with all endpoints
     */
    private EndpointEventBean makeExportsEvent() {

        return new EndpointEventBean(EEndpointEventType.REGISTERED,
                pDispatcher.getEndpoints());
    }

    /**
     * Registers the discovered endpoints
     * 
     * @param aEndpoints
     *            Registered endpoints
     * @param aSenderAddress
     *            Address of the remote isolate
     */
    private void registerEndpoints(final EndpointDescriptionBean[] aEndpoints,
            final String aSenderAddress) {

        for (final EndpointDescriptionBean endpoint : aEndpoints) {
            // Compute the import endpoint
            final ImportEndpoint importEndpoint = endpoint.toImportEndpoint();
            importEndpoint.setServer(aSenderAddress);

            // Notify the registry
            pRegistry.add(importEndpoint);
        }
    }

    /**
     * Requests the services exported by the given isolate. If no isolate is
     * given, the request is sent to all known isolates.
     * 
     * @param aIsolateUid
     *            An isolate UID (optional, can be null)
     */
    private void requestEndpoints(final String aIsolateUid) {

        // Prepare our event
        final EndpointEventBean event = makeExportsEvent();

        final ISignalSendResult rawResults;
        if (aIsolateUid == null || aIsolateUid.isEmpty()) {
            // Send the signal to all known isolates
            rawResults = pSender.sendGroup(
                    ISignalsConstants.BROADCASTER_SIGNAL_REQUEST_ENDPOINTS,
                    event, EBaseGroup.OTHERS);

        } else {
            // Send the signal to the given isolate only
            rawResults = pSender.send(
                    ISignalsConstants.BROADCASTER_SIGNAL_REQUEST_ENDPOINTS,
                    event, aIsolateUid);
        }

        if (rawResults == null) {
            // No result...
            return;
        }

        // Get successful results
        final Map<String, Object[]> sigResults = rawResults.getResults();
        if (sigResults == null || sigResults.isEmpty()) {
            // No result
            return;
        }

        // Handle all events
        for (final Entry<String, Object[]> entry : sigResults.entrySet()) {
            final String isolateUid = entry.getKey();

            // Only use the first result
            EndpointEventBean isolateEvent = null;
            for (final Object rawIsolateResult : entry.getValue()) {
                if (rawIsolateResult instanceof EndpointEventBean) {
                    isolateEvent = (EndpointEventBean) rawIsolateResult;
                    break;
                }
            }

            if (isolateEvent == null) {
                // No valid result for this isolate
                continue;
            }

            // Found a result, compute the host address of the isolate
            final String isolateNode = pDirectory.getIsolateNode(isolateUid);
            final String nodeAddress = pDirectory.getHostForNode(isolateNode);

            // Handle the event
            handleEvent(isolateEvent, nodeAddress);
        }
    }

    /**
     * Sends an endpoint event to all other isolates
     * 
     * @param aEvent
     *            The endpoint event to send
     */
    private void sendEvent(final EndpointEventBean aEvent) {

        pSender.fireGroup(ISignalsConstants.BROADCASTER_SIGNAL_REMOTE_EVENT,
                aEvent, EBaseGroup.OTHERS);
    }

    /**
     * Component validated
     */
    @Validate
    public void validate() {

        // Register to all broadcast signals
        pReceiver.registerListener(
                ISignalsConstants.PREFIX_BROADCASTER_SIGNAL_NAME
                        + ISignalsConstants.MATCH_ALL, this);

        pLogger.log(LogService.LOG_INFO, "Signals Discovery ready");

        // Request existing endpoints
        requestEndpoints(null);
    }
}
